### ISSUE TITLE
Newcheck UnnecessaryTypeArgumentsWithRecordPattern

### ISSUE THREAD

**Issue by mahfouz72** [checkstyle/checkstyle#15223]
child of #14890 

from [JEP](https://openjdk.org/jeps/440):

> If a record pattern names a generic record class but gives no type arguments (i.e., the record pattern uses a raw type) then the type arguments are always inferred. Inference works with nested record patterns. In fact it is possible to drop the type arguments in the outer record pattern as well, leading to the concise code

As of Java 21, type arguments for record patterns can be inferred by the compiler when not explicitly provided, JEP recommend to drop type arguments in record pattern as this leads to a more concise code.

Related reading from JLS : https://docs.oracle.com/javase/specs/jls/se22/html/jls-18.html#jls-18.5.5

**Note**: For compatibility reasons, type patterns do not support the implicit inference of type arguments 

---

**Config**

```XML
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="UnnecessaryTypeArgumentsWithRecordPattern"/>
    </module>
</module>
```
---
**Examples**

```Java
public class Test {
    record Box<T>(T t) { }

    void test1(Box<Box<String>> bbs) {
        if (bbs instanceof Box<Box<String>>(Box<String>(var s))) { // 2 violations
            System.out.println("String " + s);
        }
        if (bbs instanceof Box<Box<String>>(Box(var s))) { // violation
            System.out.println("String " + s);
        }
        if (bbs instanceof Box(Box(var s))) {
            System.out.println("String " + s);
        }
        // Here the compiler will infer that
        // the entire instanceof pattern is Box<Box<String>>(Box<String>(var s)),
    }

    void test2(Box obj) { // raw use
        // if (obj instanceof Box(var s)) { } this will not compile, compiler can't infer type
        // if (obj instanceof Box<String>(var s)) { } this will not compile, Box' cannot be safely cast to 'Box<String>'
        if (obj instanceof Box<?>(var s)) { // only wildcard is allowed
            System.out.println("String " + s);
        }
    }
}

```

---
Output

```
$ java -jar checkstyle-10.19.0-all.jar -c config.xml Test.java 
Starting audit...
[ERROR] D:\Test.java:5:31: Unnecessary type arguments with record pattern, They are inferred by the compiler. [UnnecessaryTypeArgumentsWithRecordPattern]
[ERROR] D:\Test.java:5:48: Unnecessary type arguments with record pattern,, They are inferred by the compiler[UnnecessaryTypeArgumentsWithRecordPattern]
[ERROR] D:\Test.java:8:31: Unnecessary type arguments with record pattern, , They are inferred by the compiler[UnnecessaryTypeArgumentsWithRecordPattern]
Audit done.
Checkstyle ends with 3 errors.
```

**Comment by rnveach** [comment]
> Here the compiler will infer that

Will it still be able to infer if `bbs` is `Object` or something similar?

````
    void test3(Object o) {
        if (o instanceof Box<Box<String>>(Box<String>(var s))) { }
````

> `if (bbs instanceof Box(Box(var s))) {`

Have you checked if there is any tools which will complain that the raw type is used? I know eclipse has an option around this for variable declarations, I am not sure what they did for patterns.

> Output
> Unnecessary type arguments with record pattern, They are inferred by the compiler.
> Unnecessary type arguments with record pattern,, They are inferred by the compiler
> Unnecessary type arguments with record pattern, , They are inferred by the compiler

You didn't specify the violation message, but if we are infer it from this section, there is no commonality. All of them have a slight difference.

Also `They` wouldn't be capitalized after a comma.

**Comment by mahfouz72** [comment]
> Will it still be able to infer if bbs is Object or something similar?

I tried it, here the compiler will infer the pattern to something like `Box(Box<?>(var s))` 

```Java
   void test1(Object bbs) {
        // if (bbs instanceof Box<Box<String>>(Box<String>(var s))) { } // Compile time error, 'Object' cannot be safely cast to 'Box<Box<String>>'
        // if (bbs instanceof Box<Box<String>>(Box(var s))) { } // Compile time error, 'Object' cannot be safely cast to 'Box<Box<String>>'
        if (bbs instanceof Box(Box(var s))) {
            System.out.println(s);
        }
    }
```
---

> Basically is there some way the compiler could infer, but we couldn't since we aren't type aware?

do we need to infer? my thoughts we just violate the usage of type arguments with record pattern because the compiler will do the work and infer them


> there is no commonality. All of them have a slight difference.
> 
> Also `They` wouldn't be capitalized after a comma.

Updated

**Comment by rnveach** [comment]
> Basically is there some way the compiler could infer, but we couldn't since we aren't type aware?
do we need to infer?

I think my original sentence is backwards. Is there a way the compiler couldn't infer but we print a violation on it since we can't double check the infer.

You showed examples that places were not "compilable" and talked about the compiler inferring, so I was worried we would violate something and it couldn't be solved by rewriting the expression where the compiler couldn't actually infer anything.

**Comment by mahfouz72** [comment]
>  Is there a way the compiler couldn't infer but we print a violation on it since we can't double check the infer.

As far as I know, there is one scenario in which the compiler could not infer. When the obj being checked is a record of a raw type, as I showed in the first post.

```
    void test2(Box obj) { // raw use
        // if (obj instanceof Box(var s)) { } this will not compile, compiler can't infer type
        // if (obj instanceof Box<String>(var s)) { } this will not compile, Box' cannot be safely cast to 'Box<String>'
        if (obj instanceof Box<?>(var s)) { // only wildcard is allowed, so we shouldn't violate type arguments if it is a wildcard
            System.out.println("String " + s);
        }
    }
```

**Comment by rnveach** [comment]
> When the obj being checked is a record of a raw type

So this check will ignore everything if they use `<?>`, even if the compiler could infer? 

Example (or something similar):
````
if (bbs instanceof Box<Box<?>>(Box<?>(var s))) {
````

If this is the case, please make this clear in the first post and add your example from the other post to the first.

> Unnecessary type arguments with record pattern, they are inferred by the compiler. 

IMO, I am not sure we need the comma and everything after it. The check description should explain why it does what it does. For example, redundant modifier check doesn't say why it is redundant. Same with our other unnecessary checks.

**Comment by mahfouz72** [comment]
> So this check will ignore everything if they use <?>, even if the compiler could infer?

Yes, flagging them means erring on the side of false positives when the compiler can't infer ( and users can't make any fix to satisfy the violation the only thing is suppression). Also ignoring them means erring on the side of false negatives. but I feel it is more reasonable to ignore them to avoid false positives that can't be fixed.

Also, I don't know if this is correct thinking but the whole point of this check is to make a nested record pattern syntax more readable and concise, I don't think adding extra `?` will affect readability in a bad way.


If we want to make things complex but 100% correct we can check by ourselves in the check if this can be inferred or not (basically do the compiler work :). ) but we will come to a point where we can't do this in some situations because we are not type-aware.


> IMO, I am not sure we need the comma and everything after it. The check description should explain why it does what it does. 

Agreed.

**Comment by rnveach** [comment]
I will give my approval for check, but I am less sure since I am unfamiliar with the syntax and the extent of the infer.

**Comment by nrmancuso** [comment]
Will we also support record patterns in switch? I do not see any examples of this.

**Comment by mahfouz72** [comment]
Yes, we will support this also, Inference can happen for records in switch. OP is updated for example

**Comment by nrmancuso** [comment]
Ok, let's do it

**Comment by gianmarcoschifone** [comment]
I'm on this

**Comment by mahfouz72** [comment]
fix is merged at https://github.com/checkstyle/checkstyle/pull/19056
